frontend/app/.server/domain/services/redis.service.ts

```diff
@@ -0,0 +1,113 @@
+import { inject, injectable } from 'inversify';
+import Redis from 'ioredis';
+import type { Logger } from 'winston';
+
+import type { ServerConfig } from '~/.server/configs';
+import { SERVICE_IDENTIFIER } from '~/.server/constants';
+import type { LogFactory } from '~/.server/factories';
+
+/**
+ * The RedisService is a service module for interacting with a Redis server. It
+ * provides methods for getting, setting, and deleting values from Redis.
+ *
+ * The getRedisService() function creates a singleton instance of the Redis
+ * client. The instance is lazy initialized as needed, so it will only be
+ * created when a client requests it. This ensures that a single connection pool
+ * is used for the entire application.
+ *
+ * Example usage:
+ *
+ *   await redisService.set('key', 'value', { EX: 600 }); // expire in 10 minutes
+ *   const value = await redisService.get('key');
+ *   await redisService.del('key');
+ *
+ * @see https://redis.io/docs/connect/clients/nodejs/
+ */
+export interface RedisService {
+  /**
+   * @see https://redis.io/commands/get/
+   */
+  get: <T>(key: string) => Promise<T | null>;
+  /**
+   * @see https://redis.io/commands/set/
+   */
+  set: (key: string, value: unknown, ttlSecs: number) => Promise<'OK'>;
+  /**
+   * @see https://redis.io/commands/del/
+   */
+  del: (key: string) => Promise<number>;
+}
+
+@injectable()
+export class RedisServiceImpl implements RedisService {
+  private readonly log: Logger;
+  private readonly redisClient: Promise<Redis>;
+
+  constructor(@inject(SERVICE_IDENTIFIER.LOG_FACTORY) logFactory: LogFactory, @inject(SERVICE_IDENTIFIER.SERVER_CONFIG) serverConfig: ServerConfig) {
+    this.log = logFactory.createLogger('RedisServiceImpl');
+    this.redisClient = serverConfig.REDIS_SENTINEL_NAME ? this.newSentinelClient(serverConfig) : this.newRedisClient(serverConfig);
```

Comment:
Can `redisClient` be injected based on `serverConfig.REDIS_SENTINEL_NAME` with a `RedisClientFactory`?

Reply:
I don't see any reason to do that. It would just clutter the project more with minimal to no benefit.

The redis service and the redis client share a single lifecycle. Managing the client outside of the only service that uses it doesn't make much sense, in my opinion.

Reply:
Ok, but it defeat the purpose of DI.

Reply:
DI is a tool and, like any tool, it can be overused. You have to use it when and where it makes sense.

Reply:
Thank you for attending my TED talk.

Reply:
The following unit tests wouldn't be required if `redisClient` was injected.

- `should create a standalone client when not using a sentinel`
- `should create a sentinel client when using a sentinel`

You're still defeating the purpose of DI.

Reply:
Cluttering the project up just to avoid writing two trivial unit tests is counter-productive.

Reply:
You're still defeating the purpose of DI by coupling `Redis` class implementation to the service.

Reply:
You're falling into the "when all you have is a hammer, everything is a nail" trap.

![image](https://github.com/user-attachments/assets/8aa7fa13-25b1-4ea2-9d58-9564988a52e0)

Reply:
Fair, can be refactored later.